File: code/python/ch06/practice/warmup_03_mystery_complexity.py
"""
Warmup 3: Mystery Complexity
==============================
Chapter 6: How Fast Is Your Code?

PROBLEM
-------
You ran a mystery function on several input sizes and recorded how many
operations it performed.  Given parallel lists of n_values and counts,
figure out the Big-O complexity class.

Classification rules (using consecutive pairs where n roughly doubles):
  - All counts are the same                  -> "O(1)"
  - Count increases by ~1 when n doubles     -> "O(log n)"
  - Count doubles when n doubles             -> "O(n)"
  - Count quadruples when n doubles          -> "O(n^2)"

INPUT FORMAT
------------
Two lines:
  Line 1: space-separated n_values (integers)
  Line 2: space-separated counts (integers)

OUTPUT FORMAT
-------------
One of: "O(1)", "O(log n)", "O(n)", "O(n^2)"

CONSTRAINTS
-----------
- At least 2 data points
- n_values are in increasing order
- Consecutive n_values roughly double (ratio close to 2)

EXAMPLES
--------
Input:
  1 10 100 1000
  5 5 5 5
Output: O(1)

Input:
  100 200 400 800
  100 200 400 800
Output: O(n)

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(n_values: list[int], counts: list[int]) -> str:
    """Classify the growth rate from observed (n, count) data points."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    n_values = list(map(int, input().split()))
    counts = list(map(int, input().split()))
    print(solve(n_values, counts))

File: code/cpp/ch06/practice/warmup_03_mystery_complexity.cpp
/*
 * Warmup 3: Mystery Complexity
 * ============================
 * Chapter 6: How Fast Is Your Code?
 *
 * PROBLEM:
 *   You are given two parallel vectors: n_values and counts. For each
 *   n_values[i], counts[i] is the number of operations observed.
 *   Determine the Big-O complexity class that best fits the data.
 *
 *   Return one of: "O(1)", "O(log n)", "O(n)", or "O(n^2)".
 *
 * EXAMPLES:
 *   solve({1,10,100,1000}, {5,5,5,5})           -> "O(1)"
 *   solve({1,2,4,8,16}, {0,1,2,3,4})            -> "O(log n)"
 *   solve({100,200,400,800}, {100,200,400,800})  -> "O(n)"
 *   solve({10,20,40,80}, {100,400,1600,6400})    -> "O(n^2)"
 *
 * CONSTRAINTS:
 *   - At least 2 data points
 *   - n_values are positive and increasing
 *   - counts are non-negative
 */

#include <iostream>
#include <string>
#include <vector>
using namespace std;

/**
 * Classifies the complexity based on observed (n, count) data points.
 */
string solve(vector<int> n_values, vector<int> counts) {
    // TODO: Replace this with your solution
    return "";
}

// -- Do not change anything below this line --------------------------
int main() {
    int k;
    cin >> k;
    vector<int> n_values(k), counts(k);
    for (int i = 0; i < k; i++) cin >> n_values[i];
    for (int i = 0; i < k; i++) cin >> counts[i];
    cout << solve(n_values, counts) << endl;
    return 0;
}
